rfd/0180-sso-as-mfa-method.md

```diff
@@ -0,0 +1,425 @@
+ http://127.0.0.1:60433/f5858c78-75e1-4f3f-b2c5-69e8e76c0ff9
+Or tap any security key
+```
+
+If the user re-authenticates with MFA, the web browser should be closed
+automatically.
+
+As in the case above, if SSO is not the preferred MFA method for the cluster,
+the browser will not be opened automatically, but the user can still click the
+link to proceed with SSO.
+
+```console
+> tsh ssh server01
+Re-authentication is required to access this node.
+Complete an auth check in your local web browser. Open it by clicking on the link:
+ http://127.0.0.1:60433/f5858c78-75e1-4f3f-b2c5-69e8e76c0ff9
+Or tap any security key
+```
+
+### Configuration
+
+#### Enable SSO connector as MFA method.
+
+IdP connectors will gain a new `allow_as_mfa_method` field with possible values
+`no`, `yes`, and `only`. 
+
+When set to `no`, the connector cannot be used for MFA checks, and vice versa
+for `yes`.
+
+When set to `only`, the connector can be used for MFA checks but cannot be used
+for login. This option is useful in cases where administrators want to set up
+a second IdP which performs a subset of the login checks. e.g. just Webauthn
+without password.
+
+Note: a split Login/MFA IdP setup requires that both IdPs are set up for the
+same list of users. If a user tries to authenticate through the MFA IdP, and
+the resulting login's username doesn't match, it will result in an error. The
+user will be prompted to add an MFA device to give them a path forward while
+administrators sort out the misconfiguration.
+
+```yaml
+kind: saml
+version: v2
+metadata:
+  name: okta
+spec:
+  allow_as_mfa_method: no | yes | only
+```
+
+#### Make SSO the only MFA method
+
+Using SSO as an MFA method enables Administrators to maintain tighter control
+over what MFA devices can be used for MFA. In some cases, it may make sense to
+disable non-SSO MFA methods to prevent users from going around registered SSO
+MFA connectors.
+
+For this use case, we will add `second_factor: sso`, which will prevent users
+from registering/using MFA devices registered through Teleport.
+
+#### Default MFA connector
+
+Cluster auth preference will also gain the `mfa_connector_name` field to set
+a preferred IdP connector for MFA checks. As detailed in the UX section, setting
+a preferred IdP connector has some beneficial UX implications.
+
+If `mfa_connector_name` is not set, but `connector_name` is set, that connector
+will be used as an MFA method if `allow_as_mfa_method` is set to `yes`.
+
+If neither of the fields above are set, Teleport will look through all
+registered connectors and return the first one with `allow_as_mfa_method` set
+to `yes` or`only`.
+
+#### Bad configuration
+
+The following two fail states should be prevented:
+- `mfa_connector_name` points to a connector with `allow_as_mfa_method = no`
+- `connector_name` points to a connector with `allow_as_mfa_method = only`
+
+These fail states will be checked on both connector update and auth preference
```

Comment:
>Administrator has a single SSO connector configured

This shouldn't be an issue. Whether the connector allows MFA, Login, or both, it should be usable for its intended purpose. There will be some logic changes necessary in the current SSO login logic - e.g. disallowing users from using an mfa-only connector explicitly with `--auth=mfa-connector` - but I think we should leave such details out of the RFD as specific implementation details.

>Administrator sets allow_as_mfa_method: yes but makes no other changes

Is the concern here that the admin would set `allow_as_mfa_method=yes|only` without updating the `cap.second_factor`?

This is probably unclear in the current draft, but SSO MFA will be accepted regardless of `second_factor`. For example, If a connector allows MFA, it will be allowed even if `second_factor=webauthn`.

`second_factor: sso` is only used to disable other MFA methods, and something like `disable_sso_mfa` would be redundant when you can just set `allow_as_mfa_method=no` on all connectors (default value).

>Without explicit configuration on the IdP side, per-session MFA is effectively neutralized, as the user's existing IdP session would satisfy MFA checks without any user interaction. Can we detect this? Could we at least warn if max_age is unset or find some other heuristic?

Agreed, my intention was to use `ForceAuthn` for SAML and find the equivalent setting for OIDC. Looks like `max_age=0` will do it.